File: Templates_and_Libraries/primes_factors.cpp
#include <iostream>
#include <vector>
#include <map>
#include <set>

using INT = long long;

std::vector<INT>
get_primes(INT MaxPrime) {
  std::vector<INT> primes;
  std::vector<bool> is_prime(MaxPrime+1, true);
  is_prime[0] = false;
  is_prime[1] = false;

  size_t i=2;
  for (; i*i<=MaxPrime; i++) {
    if (is_prime[i]) {
      primes.push_back(i);
      for (size_t j=i*i; j<=MaxPrime; j+=i) {
        is_prime[j] = false;
      }
    }
  }

  for (; i<=MaxPrime; i++) {
    if (is_prime[i]) {
      primes.push_back(i);
    }
  }

  return primes;
}

std::map<INT, INT>
get_prime_factors(INT const N) {
  std::map<INT, INT> prime_factors;

  // 空を返したほうがいい場合もある
  if (N<2) {
    prime_factors.insert(std::make_pair(N, 1));
    return prime_factors;
  }

  auto n = N;
  for (INT p=2; p*p<=N; p++) {
    while (n % p == 0) {
      prime_factors[p]++;
      n /= p;
    }
  }

  if (n > 1) prime_factors[n]++;

  return prime_factors;
}

std::set<INT> get_factors(INT n) {
  std::set<INT> ret;

  for (INT i=1; i*i<=n; i++) {
    if (n%i==0) {
      ret.insert(i);
      ret.insert(n/i);
    }
  }

  return ret;
}

int main() {
  INT n;
  std::cin >> n;

  for (auto p : get_primes(n)) {
    std::cout << p << '\n';
  }

  for (auto f : get_prime_factors(n)) {
    auto p = std::get<0>(f);
    auto k = std::get<1>(f);
    std::cout << p << ' ' << k << '\n';
  }

  for (auto f : get_factors(n)) {
    std::cout << f << '\n';
  }

  return 0;
}

File: Templates_and_Libraries/primes_factors.rs
fn get_primes(n: usize) -> Vec<usize> {
    let mut primes : Vec<usize> = vec![];
    let mut is_prime : Vec<bool> = vec![true; n+1];
    is_prime[0] = false;
    is_prime[1] = false;

    for i in 2..n+1 {
        if is_prime[i] {
            primes.push(i);
            for j in (i..).take_while(|j| i*j <= n) {
                is_prime[i*j] = false;
            }
        }
    }

    primes
}

use std::collections::BTreeMap;

fn get_prime_factors(n: usize) -> BTreeMap<usize, usize> {
    let mut prime_factors = BTreeMap::new();

    if n < 2 {
        prime_factors.insert(n, 1);
        return prime_factors;
    }

    let mut m = n;

    for f in (2..).take_while(|a| a*a <= n) {
        let mut i = 0;
        while m % f == 0 {
            i = i+1;
            m = m / f;
        }
        if i > 0 {
            prime_factors.insert(f, i);
        }
    }

    if m > 1 {
        prime_factors.insert(m, 1);
    }

    prime_factors
}

use std::collections::BTreeSet;

fn get_factors(n: usize) -> BTreeSet<usize> {
    let mut factors = BTreeSet::new();

    for i in (1..).take_while(|i| i*i <= n) {
        if n%i == 0 {
            factors.insert(i);
            factors.insert(n/i);
        }
    }

    factors
}

fn main() {
    for p in get_primes(50) {
        println!("{}", p);
    }

    for (p, k) in get_prime_factors(138) {
        println!("{}^{}", p, k);
    }

    for f in get_factors(146) {
        println!("{}", f);
    }
}
